ADR/0041-integration-service-upstreaming.md

```diff
@@ -0,0 +1,95 @@
+
+- Krunoslav Pavic
+- Andrew McNamara
+
+## Context
+
+This ADR aims to cover the proposed functionality for creating additional Artifacts as part of the Konflux build/test/release process.
+These artifacts would vary in type and would cover use cases where either the single built image or entire sets of images
+from the user's application are needed to generate one or more artifacts.
+
+## Use Cases in Detail
+
+* A team produces operands and operators. When integration service creates a Snapshot, it uses a constructor that
+  actually builds a bundle. When releasing, the team releases that single bundle artifact. That constructor also has an extract
+  operation to list all related images that are referenced, and that extraction generates the list for release service to operate on
+* A team produces a set of images with ko. The build task generates/pushes all images with ko and then produces a "build artifact"
+  OCI manifest which includes an immutable reference to each of the ko-built artifacts. When integration service creates a Snapshot,
+  it uses a constructor that doesn't need to create a new artifact. When releasing, the team releases that single artifact.
+  The constructor also has an extract operation to list all the ko-built artifacts and that extraction generates the list for release service to operate on.
+* A team produces a set of images that are just a group of artifacts that should be released together.
+  The artifacts are built via separate PipelineRuns. When integration service creates a Snapshot, it uses a constructor
+  that saves the Snapshot CR as an OCI blob. When releasing, the team releases that single artifact.
+  The constructor also has an extract operation to list all artifacts that were in the Snapshot and that extraction
+  generates the list for release service to operate on.
+
+
+## Decision
+
+A new API will be introduced which instructs the Integration service to create a Tekton pipeline which will generate the
+artifact.
+Once built, the artifact reference will be added to the Snapshot's metadata so it can be discovered during testing or release process.
+Additionally, the integration service will wait for the pipelines to finish before starting the integration testing pipelines
+in order to enable users to run tests using the new artifacts.
+
+### The ArtifactConstructor API
+
+The proposal here is to introduce the new ArtifactConstructor API which will define Tekton pipelines that will generate individual artifacts.
+These pipelines would be executed after a Snapshot is generated and before the integration test pipelines start.
+
+When each artifact generation pipelineRun is completed, the reference to the artifact is extracted and added to the Snapshot.
+After that, the integration testing pipelines castart, and since the artifacts are referenced by the Snapshot,
+it is possible to fetch them and run tests on them.
+
+```yaml
+apiVersion: appstudio.redhat.com/v2alpha1
+kind: ArtifactConstructor
```

Comment:
This will result in pipelines run in users' namespaces without the use of PAC. We would lose the ability to interact with PipelineRuns as users might normally be used to (i.e. re-triggering them). How could we ensure that we still provide users with that necessary functionality?

Reply:
The integration pipelines are also run without the use of PaC, and there is already a re-triggering functionality available for them, would that be enough to cover the use cases?

Reply:
It might help in some use cases. I don't know how narrowly scoped those are.

Another use case that we need to consider is if the artifact-generating repository reference were updated, we would need to be able to test the new images/build process.

I don't know what the best way to do that is. Would it make sense to still add this Component and then designate some component's repo/configuration as the constructor argument? This feels a little wonky to me.